Update default language when bulk uploading translations

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-14776

### Problem
When attempting to create an app build, an error saying that multiple default languages are set causes the build to fail. This error is due to code in uploading bulk translations.

An app can have multiple languages configured, but only one is marked as the default language. This is set in two places (not ideal):
1. On the application doc directly where the first language in the list of supported languages is the default
2. On each form within the application where a `default` attribute is set on the relevant translation node

When updating the app's default language, only the app doc is updated, **not each form**. The forms are updated once a new build is created. One solution to this problem could be to update each form along with the app when changing the default language, but that can be a very costly operation if there are lots of forms, since the form needs to be rendered in order to modify it.

So we are in a situation where _most_ of the time, an app's newly default language isn't reflected in forms until a new build is created, with the exception of an edge case in bulk translation uploads. This is because when uploading translations, the app's default language is referenced if **new translation nodes need to be created**. This is to ensure the new nodes have `default` set correctly.

The problem is if the default language was modified on the app, but not yet in the forms (a new build hasn't been created), the bulk translation upload process doesn't check to see if any `default` attributes need to be removed from existing translation nodes, but only checks if `default` attributes need to be added to new translation nodes. This is how we can end up in a scenario where there are multiple translation nodes with default set.

### Solution

As mentioned above, one solution could be to update each form when updating the default language of an app, but that has the potential to be a very costly action that would not lend itself well to a synchronous operation.

Since we are already iterating over translation nodes in this process, and this is the process that causes the issues in the first place, the easiest solution is just fix this process to not introduce any potential build issues. We did think about removing the entire concept of default languages from the bulk translation upload process as well, but determined we could not safely do so because of how we use existing translation nodes as the template for the new node.

So we settled on this solution to just ensure the default attribute is set on each translation node once the new nodes are created as part of the bulk translation upload process.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added a regression test to ensure all translation elements are updated to reflect the current default language.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi.atlassian.net/browse/QA-7084

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Update default language for every translation node

When uploading bulk translations, ensure translation nodes are up to
date with the app's language config.

nospame commented on corehq/apps/translations/app_translations/upload_form.py:
```diff
@@ -142,13 +143,14 @@ def _add_missing_translation_elements_to_itext(self, template_translation_el):
-                    new_trans_el.set('default', '')
                 self.itext.xml.append(new_trans_el)
 
+    def _modify_default_lang_if_needed(self, lang, trans_el):
```
Maybe it's weird to bring up now since we paired on this, but I don't love the name, especially since we're always adding or removing the `default` attribute whether a change is needed or not. 

What would you think about something in the realm of `_set_translation_node_default_lang` (wordy but gets to the core of what's happening) or `_update_node_default_attr` (concise and more literally what's happening)?

gherceg replied:
Hah you're allowed to voice opinions after we paired. I agree that we indiscriminately attempt to update the node, but that is an implementation detail of the method that I don't think _has_ to be reflected in the method name. The intent is still that if the node already has the 'default' attr set, nothing changes (and vice versa).

So I guess the question you're getting at is do we get value out of appending `if_needed` or `if_necessary` to this method name, or would it only cause confusion once a dev looks at the implementation which I could imagine could be true. To take one out of Matt's playbook, if we try to say in plain english what the intent of this method is, what do we get? I would say we are making sure that the current node is up to date with the app's default language. So maybe something like `def _update_default_language(self, node)`? I'm in favor of changing the method to look up the language on the node as well to simplify things.

nospame replied:
I'm not opposed to `if_needed`, I think I'm just wanting it to be _really clear_ where this is updating the default and based on what criteria, since the multiple places we keep a "default" seem to be part of what made investigating more difficult. I think argument naming might actually help a lot here. With all this in mind I think my preference would be something close to `def _set_language_as_default_if_needed(self, node_lang, node)`

gherceg replied:
Yeah realistically I think a comment is worthwhile since this is a bit of a workaround/hack anyway, and we won't be able to capture all of that context in just a method name. I know this contradicts my statement about abstracting implementations away, but I think I'm content with `_update_default_attr_if_needed(node, node_lang)`.

c35a5dc31de08757b8e7e0391834f560da03b9bb

gherceg replied:
Ok actually I think most of our issues were because where we were calling this method, not the method name. I changed this in 1c503f4881f85ac05f0a675ed47ae670841ce477 so that it is hopefully more intuitive what is happening and why.

gherceg committed:
Rename method and add comment explaining why

gherceg committed:
Still update default attr after creating new node

For the sake of preserving behavior, still run this code when creating a
new code.

gherceg committed:
Create _setup_translation_nodes method

It makes sense for all of this logic to live in the same place

gherceg committed:
Naming updates for _get_template_translation_el

gherceg committed:
Rename node to element

Also, use _update_or_create instead of _setup

gherceg committed:
Merge branch 'master' into gh/bulk-translations/update-default-lang

gherceg committed:
Missed some node -> element renames

nospame commented on corehq/apps/translations/app_translations/upload_form.py:
```diff
@@ -117,37 +116,54 @@ def update(self, rows):
-                else:
-                    new_trans_el.set('default', '')
-                self.itext.xml.append(new_trans_el)
+        raise Exception(_("Form has no translation element present to be used as a template."))
```
I know you/we didn't add this, but is there a more specific exception that could be raised here?

gherceg replied:
Sure. I checked the call stack and don't see anywhere that handles XFormExceptions specifically so I don't anticipate any behavior to change with this change and agree it is generally a good thing to do. 9dc05ba2c6c8ece942392597c063f663b2dd7d78

gherceg committed:
Raise XFormException instead of general Exception

gherceg committed:
Test translation elements are updated

Ensure all translation elements are updated to reflect the new default
language

gherceg commented on corehq/apps/translations/app_translations/upload_form.py:
```diff
@@ -77,7 +77,7 @@ def update(self, rows):
         # Setup
         rows = get_unicode_dicts(rows)
-        self._setup_translation_nodes()
+        self._update_or_create_translation_elements()
```
Evan was in favor of element because it was more specific than node, since everything in an xml doc is a node. Agreed that it is generally a losing battle due to the lack of convention.

gherceg replied:
That sounds like I'm blaming Evan for that decision which I don't mean to 😅 I think element > node > el, but because of the lack of convention, the difference isn't important in the grand scheme of things.

gherceg replied:
Ok actually Jenny pointed out a gap in my knowledge, which is that I don't know that there is actually a meaningful difference between element and node. @nospame are you aware of elements being a specific xml convention that makes them "more specific" than nodes?

nospame replied:
My pretty surface-level knowledge is that "node" is specific to XML, but basically describes anything selectable -- attributes, elements, even the document itself. In this context it seemed like node was the superset and that "element" would be more specific.

orangejenny replied:
Attributes aren't nodes. I had thought element and node were exactly equivalent but found [this](https://docs.oracle.com/en/cloud/saas/marketing/responsys-develop/RPL/XMLTerminology.htm) today saying the document doesn't count as an element. So element is a tiny bit more specific.

nospame replied:
I had looked at [this](https://www.w3schools.com/xml/dom_nodes.asp) saying attributes are a type of node. In any case, I think we're in agreement that element is _slightly_ more specific in this case.

orangejenny replied:
Huh, maybe I'm wrong. I see docs saying both, and also some mention of "attribute nodes" 🤷‍♀️

gherceg commented:
Requested QA: https://dimagi.atlassian.net/browse/QA-7084
